enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,187 @@
+
+The data provided by the user will need to have the following structure:
+```
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```
+master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+#### API Open Questions
+
+* How do we map an individual machine's keyfiles to the machine? Hostname was previously suggested because it was already an available label in NNCP, but if we're not using NNCP then there might be a better way.
+
+* If we're using RHCOS functionality, do we still need the custom image build discussed below? Or can we just pass the keyfiles as part of the deployment process?
+
+* What is the best way to encode the data above into secret(s)?
+
+### Processing user configuration
+
+#### Deployment of the controlplane hosts via terraform
+
+TODO - we'll have to match the keyfiles to the master Machines, then generate image/images for deployment via terraform/ironic
+
+#### Deployment of compute/worker hosts
+
+TODO - this will rely on metal3 image-building proposal https://github.com/metal3-io/metal3-docs/pull/183
+
+## Alternatives
+
+### Use Kubernetes-NMState NodeNetworkConfigurationPolicy custom resources
+
+If we were able to install the NNCP CRD at day-1 then we could use that as the configuration interface. This has the advantage of matching the configuration syntax and objects used for day-2 network configuration via the operator.
+
+However, we have been unable to get agreement on a method to deliver the operator on day-1, which makes this problematic. It's possible we could implement some way to pull just the NNCP CRD from the operator payload and isntall it on day-1, but at this point we likely don't have time to do that and it's not clear we even want to if the operator can't be installed on day-1 too. This would essentially orphan the CRs after the initial deployment completes.
+
+### Provide raw nmstate data
+
+We could expose an interface to the user which accepts raw nmstate data (not a CR wrapper).  This would mean a common DSL to the NodeNetworkConfigurationPolicy resource, but not directly reusing that API, so there is risk of drift between the two APIs.
+
+The user would have to pass the nmstate data via the install-config, and also as a Secret for use on scale-up, then also define a NNCP resource if day-2 management of the configuration is required (which could be risky if the configuration isn't exactly the same...)
```

Comment:
Couldn't the installer set up all of these?

Or the CAPBM could create the Secret to use for day-1 automatically from a NNCP resource that is to be used for day-2.

Reply:
The problem is ref #747 that the NNCP CRD provided via kubernetes-nmstate doesn't exist at the point of install - it's currently only applied if/when you install the operator day-2 from OLM.

We could look at making CAPBM look for such a resource in future though, then fall back to using secrets (which might resolve some of the concerns around day1/day2 interface overlap is we can make it clear to users which API they should interact with)

Reply:
Ah, I was missing that context.
I wonder if the objections to #747 could be overcome by saying we'll enable the operator at install time whenever the user specifically provides network configuration in the installer, rather than always on particular platforms.

Reply:
@zaneb perhaps - the impression I got from discussion with @dhellmann and @derekwaynecarr is that we don't want knmstate in the release payload at all, because it's not relevant to all platforms (although arguably it could be e.g to support secondary interfaces for OCS) (there are also some unresolved questions around support status and ownership to deal with).

But if we're going to block core-platform functions like network configuration from the release payload long term, it probably does mean we need some install-time way to enable optional dependencies from OLM or other sources.

Reply:
@hardys I would argue that we need network configuration on all platforms. OVN requires OVS network configuration on every single platform, which today is done by the MCO ovs-configuration bash script (creating network manager config) and would be much better served by nmstate or something more robust.

Reply:
@trozet I completely agree, but since we've not managed to achieve consensus on how to do that (ref #747) the scope of this PR is limited to using the interfaces which are currently supported (e.g keyfiles).